tests/integration_test/ota_pal/aws_test_ota_pal.c

```diff
@@ -0,0 +1,1137 @@
+{
+    OtaPalStatus_t xOtaStatus;
+    OtaImageState_t eImageState = OtaImageStateUnknown;
+    int16_t blocksWritten;
+
+    /* Create a local file again using the PAL. */
+    xOtaFile.pFilePath = ( uint8_t * ) otatestpalFIRMWARE_FILE;
+    xOtaFile.fileSize = sizeof( ucDummyData );
+
+    xOtaStatus = otaPal_CreateFileForRx( &xOtaFile );
+    TEST_ASSERT_EQUAL( OtaPalSuccess, OTA_PAL_MAIN_ERR( xOtaStatus ) );
+
+    /* We still want to close the file if the test fails. */
+    if( TEST_PROTECT() )
+    {
+        /* Write data to the file. */
+        blocksWritten = otaPal_WriteBlock( &xOtaFile,
+                                           0,
+                                           ucDummyData,
+                                           sizeof( ucDummyData ) );
+        TEST_ASSERT_EQUAL( sizeof( ucDummyData ), blocksWritten );
+
+        /* Set the image state. */
+        eImageState = OtaImageStateTesting;
+        xOtaStatus = otaPal_SetPlatformImageState( &xOtaFile, eImageState );
+        TEST_ASSERT_EQUAL_INT( OtaPalSuccess, OTA_PAL_MAIN_ERR( xOtaStatus ) );
+
+        /* Verify that image state was saved correctly. */
+
+        /* [**]All platforms need a reboot of a successfully close image in order to return
+         * OtaPalImageStatePendingCommit from otaPal_GetPlatformImageState(). So this cannot be tested.
+         */
+
+        /*ePalImageState = otaPal_GetPlatformImageState();
+         * TEST_ASSERT_EQUAL_INT( OtaPalImageStatePendingCommit, ePalImageState ); */
+    }
+}
+
+/**
+ * @brief Set an invalid platform image state exceeding the range and verify success.
+ */
+TEST( Full_OTA_PAL, otaPal_SetPlatformImageState_InvalidImageState )
+{
+    OtaPalStatus_t xOtaStatus;
+    OtaImageState_t eImageState = OtaImageStateUnknown;
+    int16_t blocksWritten;
+
+    /* Create a local file again using the PAL. */
+    xOtaFile.pFilePath = ( uint8_t * ) otatestpalFIRMWARE_FILE;
+    xOtaFile.fileSize = sizeof( ucDummyData );
+
+    xOtaStatus = otaPal_CreateFileForRx( &xOtaFile );
+    TEST_ASSERT_EQUAL( OtaPalSuccess, OTA_PAL_MAIN_ERR( xOtaStatus ) );
+
+    /* We still want to close the file if the test fails. */
+    if( TEST_PROTECT() )
+    {
+        /* Write data to the file. */
+        blocksWritten = otaPal_WriteBlock( &xOtaFile,
+                                           0,
+                                           ucDummyData,
+                                           sizeof( ucDummyData ) );
+        TEST_ASSERT_EQUAL( sizeof( ucDummyData ), blocksWritten );
+
+        /* Try to set an invalid image state. */
+        eImageState = ( OtaImageState_t ) ( OtaLastImageState + 1 );
+        xOtaStatus = otaPal_SetPlatformImageState( &xOtaFile, eImageState );
+        TEST_ASSERT_EQUAL( OtaPalBadImageState, OTA_PAL_MAIN_ERR( xOtaStatus ) );
+    }
+}
+
+/**
+ * @brief Set the image state to unknown and verify a failure.
+ */
+TEST( Full_OTA_PAL, otaPal_SetPlatformImageState_UnknownImageState )
```

Comment:
Most of the SetPlatformImageState tests can be combined to one:
```C
prvSetAndVerifyPlatformState(ImageState, ExpectedPalStatus)
{
    OtaPalStatus_t xOtaStatus;
    OtaImageState_t eImageState = ImageState;
    int16_t blocksWritten;
  .....
  
        xOtaStatus = otaPal_SetPlatformImageState( &xOtaFile, eImageState );
        TEST_ASSERT_EQUAL( ExpectedPalStatus, OTA_PAL_MAIN_ERR( xOtaStatus ) );
}

TEST( Full_OTA_PAL, otaPal_SetPlatformImageState_UnknownImageState )
{
/*Test for unknown state*/
prvSetAndVerifyPlatformState(OtaImageStateUnknown, OtaPalSuccess)
....
}

Reply:
It's realistic to have a device fail a subset of the SetPlatformImageState tests so I would like to keep these separate for now.

Reply:
You could still keep the tests separate and have them re-use the testing code.

Reply:
^Resolved this offline.